README.md
# dbt-oracle

dbt "adapters" are responsible for adapting dbt's functionality to a given database. `dbt-oracle` implements dbt functionalities for the Oracle database. To learn more about building adapters, check
https://docs.getdbt.com/docs/contributing/building-a-new-adapter

> Prior to version 1.0.0, dbt-oracle was created and maintained by [Indicium](https://indicium.tech/) on [their GitHub repo](https://github.com/techindicium/dbt-oracle). Contributors in this repo are credited for laying the groundwork and maintaining the adapter till version 0.4.3.
From version 1.0.0, dbt-oracle is maintained and distributed by Oracle.

## What is dbt? <a name='what-is-dbt'></a>

dbt does the T in ELT (Extract, Load, Transform). To work with dbt you need a copy of your data already loaded in your warehouse.

### dbt features <a name='core-concepts'></a>
- With dbt, you can express all transforms with SQL select
  - Different materialization strategies.
    - view
    - table
    - incremental; selective rebuild for new rows
    - ephemeral; Model 1 interpolated into Model 2 as a Common Table Expression (CTE)
  - No need to write boilerplate code
    - All code to create table or views is generated using macros.
  - Idempotence; rerun models
    - If your source data were to stop updating, successive runs of your transformations would still result in the same tables and views in your warehouse.
    - If your production deployment of your transformations were interrupted, the next run of the transformations would result in the same tables and views as if the deployment had not been interrupted.
    - If you manually triggered transformations between scheduled runs, the scheduled run would result in the same tables and views as if the manual runs had not been triggered.
  - All transformation code is accessible and can be version controlled.
- Dependency resolution
  - Use of ref() function ``select * from {{ ref('MODEL_NAME')}}``
  - dbt automatically resolves dependencies in between models and builds a Directed Acyclic Graph (DAG).
    Each path in the DAG can be independently executed using multiple threads.
  - Interpolates the name of database schema
- Includes a built-in testing framework to ensure model accuracy
  - not null
  - unique
  - contains accepted values
  - relationships
  - custom tests
- Generate documentation for your project and render it as a website.
- Use macros to write reusable SQL

### An example <a name='an-example'></a>

dbt model
```sql
--models/sales_internet_channel.sql
{{ config(materialized='table') }}
WITH sales_internet AS (
       SELECT * FROM {{ source('sh_database', 'sales') }}
       WHERE channel_id = 4 )
SELECT * FROM sales_internet
```
dbt compiles the above SQL template to run the below DDL statement.
```sql
CREATE TABLE dbt_test.sales_internet_channel AS
WITH sales_internet AS (
         SELECT * from sh.sales
         WHERE channel_id = 4 )
SELECT * FROM sales_internet
```

For dbt documentation, refer https://docs.getdbt.com/docs/introduction

## Installation <a name='install'></a>

dbt-oracle can be installed via the Python Package Index (PyPI) using pip

`pip install -U dbt-oracle`

### Support <a name='support'></a>

dbt-oracle will provide support for the following

- Python versions 3.6, 3.7, 3.8 and 3.9
- Autonomous Database versions 19c and 21c
- OS
  - Linux
  - MacOS
  - Windows

### Core dependencies <a name='core-dependencies'></a>
dbt-oracle requires the following 3 python packages.

`dbt-core`

  - Open source framework for data transformation
  - Jinja Templating and core SQL compilation logic
  - Latest version of dbt-core is preferred; From version 1.0.0, dbt-core supports Python 3.7 or higher
  - For Python 3.6, pip will fallback to version 0.21.1 of dbt-core

`cx-Oracle`
 - Python driver for Oracle database
 - Oracle client libraries should be installed on the system. For details check, https://cx-oracle.readthedocs.io/en/latest/user_guide/installation.html

`dataclasses; python_version < '3.7'`
 - dataclasses package was introduced in the standard Python library from Python 3.7. This is conditional dependency and required only for Python 3.6
 
## Getting Started <a name='getting-started'></a>

Create a dbt project for oracle database using the `dbt init` command. The init command is interactive and will help you get started with a new project.

`dbt init` will:

* ask you the name of the project
* ask you the database adapter you are using i.e. oracle
* prompt to specify necessary connection details

This example shows initialization of test project `dbt_oracle_test_project`

```text
>> dbt init

Running with dbt=1.0.4
Enter a name for your project (letters, digits, underscore): dbt_oracle_test_project
Which database would you like to use?
[1] oracle
  Enter a number: 1
  protocol (tcp or tcps) [tcps]: 
  host (adb.<oci-region>.oraclecloud.com) [{{ env_var('DBT_ORACLE_HOST') }}]: 
  port [1522]: 
  user [{{ env_var('DBT_ORACLE_USER') }}]: 
  password [{{ env_var('DBT_ORACLE_PASSWORD') }}]: 
  service (service name in tnsnames.ora) [{{ env_var('DBT_ORACLE_SERVICE') }}]: 
  schema (database schema in which dbt objects should be created) [{{ env_var('DBT_ORACLE_SCHEMA') }}]: 
  threads (1 or more) [1]: 4
Profile dbt_oracle_test_project written to ~/.dbt/profiles.yml using target's profile_template.yml and your supplied values. Run 'dbt debug' to validate the connection.
Your new dbt project "dbt_oracle_test_project" was created!

```

Then dbt init command will:

1. Create the following folder with project name and sample files to get you started
    ```text
     ├── README.md
     ├── analyses
     ├── dbt_project.yml
     ├── macros
     ├── models
     │   └── example
     ├── seeds
     ├── snapshots
     └── tests
    ```
2. Create a connection profile on your local machine. The default location is `~/.dbt/profiles.yml`

    Next step, [configure connection][1] related parameters and test if dbt connection works using dbt debug command
    
    ```text
    >> dbt debug
     
    os info: macOS-11.6-x86_64-i386-64bit
       Using profiles.yml file at ~/.dbt/profiles.yml
       Using dbt_project.yml file at /dbt_oracle_test_project/dbt_project.yml
       Configuration:
        profiles.yml file [OK found and valid]
        dbt_project.yml file [OK found and valid]
       Required dependencies:
       - git [OK found]
       Connection:
        user: ***
        schema: ***
        protocol: tcps
        host: adb.us-ashburn-1.oraclecloud.com
        port: 1522
        service: <service_name>_high.adb.oraclecloud.com
        connection_string: None
        shardingkey: []
        supershardingkey: []
        cclass: None
        purity: None
        Connection test: [OK connection ok]
    
       All checks passed!
    ```

## Documentation [TODO] <a name='documentation-todo'></a>
Link to the homepage - https://oracle.github.io/dbt-oracle

Link to documentation - https://dbt-oracle.readthedocs.io

## Contributing <a name='contributing'></a>
This project welcomes contributions from the community. Before submitting a pull request, please review our [contribution guide][2].

## Security <a name='security'></a>
Please consult the [security guide][3] for our responsible security vulnerability disclosure process.

## License <a name='license'></a>
dbt-oracle is licensed under Apache 2.0 License which you can find [here][4]

[1]: https://github.com/oracle/dbt-oracle/blob/main/dbt_adbs_test_project/profiles.yml
[2]: https://github.com/oracle/dbt-oracle/blob/main/CONTRIBUTING.md
[3]: https://github.com/oracle/dbt-oracle/blob/main/SECURITY.md
[4]: https://github.com/oracle/dbt-oracle/blob/main/LICENSE.txt


HISTORY.md
# History

## v0.1.0 (2020-06-02)

- First release on PyPI.

## v0.2.0 (2020-06-02)

- Added full macro implementation. Thanks Fabrice Etanchaud

## v0.4.3 (2021-06-28)

- Last stable release of [dbt-oracle](https://github.com/techindicium/dbt-oracle) by Techindicium

## v1.0.0 (2022-05-04)

- First release of vendor-built adapter by Oracle. 
- This version supports Python 3.6, 3.7, 3.8 and 3.9


dbt/adapters/oracle/connections.py
"""
Copyright (c) 2022, Oracle and/or its affiliates.
Copyright (c) 2020, Vitor Avancini

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
"""
from typing import List, Optional, Tuple, Any
from contextlib import contextmanager
from dataclasses import dataclass, field
import enum
import time

import cx_Oracle
from cx_Oracle import Connection

import dbt.exceptions
from dbt.helper_types import Port

from dbt.adapters.base import Credentials
from dbt.adapters.sql import SQLConnectionManager
from dbt.contracts.connection import AdapterResponse

# dbt.events was introduced in dbt-core==1.0.0 which only supports Python 3.7
# For Python 3.6, we fallback to the old logger in dbt.logger.
try:
    from dbt.events import AdapterLogger
    logger = AdapterLogger("oracle")
except ImportError:
    from dbt.logger import GLOBAL_LOGGER as logger



class OracleConnectionMethod(enum.Enum):
    HOST = 1
    TNS = 2
    CONNECTION_STRING = 3


@dataclass(init=False)
class OracleAdapterCredentials(Credentials):
    """Collect Oracle credentials

    An OracleConnectionMethod is inferred from the combination
    of parameters profiled in the profile.
    """
    # Mandatory required arguments.
    user: Optional[str] = None
    password: Optional[str] = None
    # Database is not needed when initiating an Oracle DB connecting
    database: Optional[str] = None
    schema: Optional[str] = None

    # OracleConnectionMethod.TNS
    tns_name: Optional[str] = None

    # OracleConnectionMethod.HOST
    protocol: Optional[str] = None
    host: Optional[str] = None
    port: Optional[Port] = None
    service: Optional[str] = None

    # OracleConnectionMethod.CONNECTION_STRING
    connection_string: Optional[str] = None

    # shardingkey and supershardingkey is a list
    shardingkey: Optional[List[str]] = field(default_factory=list)
    supershardingkey: Optional[List[str]] = field(default_factory=list)

    # Database Resident Connection Pooling (DRCP)
    cclass: Optional[str] = None
    purity: Optional[str] = None

    _ALIASES = {
        'pass': 'password',
    }

    def __init__(self, **kwargs):
        for k, v in kwargs.items():
            setattr(self, k, v)
            self.database = None

    def __post_init__(self):
        # In Oracle the userenv DB_NAME (database) is not needed when initiating a connection
        if self.database is not None:
            raise dbt.exceptions.RuntimeException(
                f'    database: {self.database} \n'
                f'With Oracle DB the database property must not be set'
            )
        self.database = None

    @property
    def type(self):
        return 'oracle'

    @property
    def unique_field(self):
        return self.user

    def _connection_keys(self) -> Tuple[str]:
        """
        List of keys to display in the `dbt debug` output. Omit password.
        """
        return (
            'user', 'schema',
            'protocol', 'host', 'port', 'tns_name',
            'service', 'connection_string',
            'shardingkey', 'supershardingkey',
            'cclass', 'purity'
        )

    def connection_method(self) -> OracleConnectionMethod:
        """Return an OracleConnecitonMethod inferred from the configuration"""
        if self.connection_string:
            return OracleConnectionMethod.CONNECTION_STRING
        elif self.host:
            return OracleConnectionMethod.HOST
        else:
            return OracleConnectionMethod.TNS

    def get_dsn(self) -> str:
        """Create dsn for cx_Oracle for either any connection method

        See https://cx-oracle.readthedocs.io/en/latest/user_guide/connection_handling.html"""

        method = self.connection_method()
        if method == OracleConnectionMethod.TNS:
            return self.tns_name
        if method == OracleConnectionMethod.CONNECTION_STRING:
            return self.connection_string

        # Assume host connection method OracleConnectionMethod.HOST
        return f'{self.protocol}://{self.host}:{self.port}/{self.service}'


class OracleAdapterConnectionManager(SQLConnectionManager):
    TYPE = 'oracle'

    @classmethod
    def open(cls, connection):
        if connection.state == 'open':
            logger.debug('Connection is already open, skipping open.')
            return connection
        credentials = cls.get_credentials(connection.credentials)
        method = credentials.connection_method()
        dsn = credentials.get_dsn()

        logger.debug(f"Attempting to connect using Oracle method: '{method}' "
                     f"and dsn: '{dsn}'")

        conn_config = {
            'user': credentials.user,
            'password': credentials.password,
            'dsn': dsn
        }

        if credentials.shardingkey:
            conn_config['shardingkey'] = credentials.shardingkey

        if credentials.supershardingkey:
            conn_config['supershardingkey'] = credentials.supershardingkey

        if credentials.cclass:
            conn_config['cclass'] = credentials.cclass

        if credentials.purity:
            purity = credentials.purity.lower()
            if credentials.purity == 'new':
                conn_config['purity'] = cx_Oracle.ATTR_PURITY_NEW
            elif purity == 'self':
                conn_config['purity'] = cx_Oracle.ATTR_PURITY_SELF
            elif purity == 'default':
                conn_config['purity'] = cx_Oracle.ATTR_PURITY_DEFAULT

        try:
            handle = cx_Oracle.connect(**conn_config)
            connection.handle = handle
            connection.state = 'open'
        except cx_Oracle.DatabaseError as e:
            logger.info(f"Got an error when attempting to open an Oracle "
                        f"connection: '{e}'")
            connection.handle = None
            connection.state = 'fail'

            raise dbt.exceptions.FailedToConnectException(str(e))

        return connection

    @classmethod
    def cancel(cls, connection):
        connection_name = connection.name
        oracle_connection = connection.handle

        logger.info("Cancelling query '{}' ".format(connection_name))

        try:
            Connection.close(oracle_connection)
        except Exception as e:
            logger.error('Error closing connection for cancel request')
            raise Exception(str(e))

        logger.info("Canceled query '{}'".format(connection_name))

    @classmethod
    def get_status(cls, cursor):
        # Do oracle cx has something for this? could not find it
        return 'OK'

    @classmethod
    def get_response(cls, cursor):
        # number of rows fetched for a SELECT statement or
        # have been affected by INSERT, UPDATE, DELETE and MERGE statements
        return AdapterResponse(rows_affected=cursor.rowcount,
                               _message='OK')

    @contextmanager
    def exception_handler(self, sql):
        try:
            yield

        except cx_Oracle.DatabaseError as e:
            logger.info('Oracle error: {}'.format(str(e)))

            try:
                # attempt to release the connection
                self.release()
            except cx_Oracle.Error:
                logger.info("Failed to release connection!")
                pass

            raise dbt.exceptions.DatabaseException(str(e).strip()) from e

        except Exception as e:
            logger.info("Rolling back transaction.")
            self.release()
            if isinstance(e, dbt.exceptions.RuntimeException):
                # during a sql query, an internal to dbt exception was raised.
                # this sounds a lot like a signal handler and probably has
                # useful information, so raise it without modification.
                raise e

            raise dbt.exceptions.RuntimeException(e) from e

    @classmethod
    def get_credentials(cls, credentials):
        return credentials

    def add_query(
        self,
        sql: str,
        auto_begin: bool = True,
        bindings: Optional[Any] = {},
        abridge_sql_log: bool = False
    ) -> Tuple[Connection, Any]:
        connection = self.get_thread_connection()
        if auto_begin and connection.transaction_open is False:
            self.begin()

        logger.debug('Using {} connection "{}".'
                     .format(self.TYPE, connection.name))

        with self.exception_handler(sql):
            if abridge_sql_log:
                log_sql = '{}...'.format(sql[:512])
            else:
                log_sql = sql

            logger.debug(f'On {connection.name}: f{log_sql}')
            pre = time.time()

            cursor = connection.handle.cursor()
            cursor.execute(sql, bindings)
            connection.handle.commit()
            logger.debug(f"SQL status: {self.get_status(cursor)} in {(time.time() - pre)} seconds")
            return connection, cursor

    def add_begin_query(self):
        connection = self.get_thread_connection()
        cursor = connection.handle.cursor
        return connection, cursor
